Add Domain/Tags.cpp, reduce includes in Domain/Tags.hpp

nilsdeppe opened the pull request:
## Proposed changes

- Add a `Domain/Tags.cpp` to move the calls to DomainCreator into there. This means we no longer need to include `DomainCreator.hpp` in `Tags.hpp`, which results in rebuilding ~60 targets instead of ~180 when changing one of the domain creator header files.
- Add dim, target_frame, source_frame to map tags so that we don't need includes for the ElementMap (and CoordinateMap in the future)
- Replace unnecessary includes in `Domain/Tags.hpp` with forward decls.

Depends on:
- [x] #2026 

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Make metavariables to create_from_options optional

nilsdeppe committed:
Add dim, target_frame, source_frame to map tags

nilsdeppe committed:
Reduce includes in Domain/Tags.hpp

moxcodes commented on src/Domain/Tags.cpp:
```diff
@@ -0,0 +1,53 @@
+// Distributed under the MIT License.
```
The commit message has some grammar trouble that I'm having trouble parsing.

kidder commented on src/Evolution/Executables/Burgers/EvolveBurgers.hpp:
```diff
@@ -5,6 +5,7 @@
 
 #include <vector>
 
+#include "Domain/Creators/DomainCreator.hpp"
```
Shouldn't this include (and in all the other executables) be somewhere else?

nilsdeppe replied:
I guess in a way it could be in `Domain/OptionsTags.hpp` but then it gets sucked in everywhere again. I would say `ParallelAlgorithms/DG/InitializeDomain.hpp` is reasonable. Thoughts?

nilsdeppe replied:
Actually, it's `DgElementArray.hpp` that adds it to the cache. I'll add the includes there.

nilsdeppe committed:
Move DomainCreator calls to Domain/Tags.cpp

Since a lot of files include Domain/Tags.hpp, not including all the domain
creators into Tags.hpp reduces what needs to be recompiled when the domain
creators are updated. This reduces the number of targets that need to be rebuilt
when any domain creator has its header file changed by ~120.
